src/app/server/java/AndroidAppServerWrapper.cpp

```diff
@@ -70,3 +67,6 @@ void ChipAndroidAppShutdown(void)
     chip::Server::GetInstance().Shutdown();
     chip::Platform::MemoryShutdown();
 }
+
+// this is only for compile, the real impl of this function is in the example project.
+void __attribute__((weak)) ApplicationInit() {}
```

Comment:
I had a  similar question in another PR: can we avoid attribute weak functions and just fail at compile time if the function is not defined?

Reply:
Yes, since another PR is closed, I changed it here.
This part of code is compiled into a so, so no matter the function is defined or not ( at example ), the compile will fail.

Reply:
Can we provide an application init callback through the java code instead?

Reply:
Please explain the order of the loads and calls. I would generally try hard to avoid weak linkage especially in dynamic linking ... I am actually unsure how attribute weak and .so is supposed to interact.

I would find it more natural to have some sort of callback. From what I see right now:
    - `JNI_METHOD(jboolean, startApp)(JNIEnv * env, jobject self)` calls ChipAndroidAppInit
    - we want `ChipAndroidAppInit` to call ApplicationInit
    
why can't `startApp` setup the necessary callbacsk (e.g. have a callback passed to it)?

Reply:
Because startApp is a java native function, it can only set up java callback but we need native init.

ApplicationInit is simliary with Linux platform, examples/platform/linux/AppMain.cpp calling this API and examples/all-clusters-app/linux/main-common.cpp using that.
They did not use weak function because examples/platform/linux/ is a source_set in gn but android it is a dynamic link library.

The call order is: java app  --> load libTvApp with ApplicationInit --> included libCHIPAppServerJni.a -> call ChipAppServer()::startApp() in libCHIPAppServerJni.a --> ChipAndroidAppInit in libCHIPAppServerJni.a --> ApplicationInit in libTvApp.so

Reply:
fixed, the weak function now is removed and fails at compile time if the function is not defined?

Reply:
What I am thinking is that java generally has bi-directional support: java can call native code and native code can call java callbacks.

If the above is correct, which language is what version should not matter:
  - java code can get a native pointer to the "ApplicationInit" method
  - java code can pass in a callback to "startApp" ... maybe implement a thing on "self" that is a callback so you can "getMethod" and call it if exists or some other way.
 
I would like the code to make call paths clear: if you need something called at startup, a pointer to such a callback/delegate should be passed in. Compile time seems stranger to me as java dynamically links jni libraries, so I am unclear how attribute weak is guaranteed to behave in a .so environment. I tried to find documentation that explains how these are suposed to work and could not find any - if you go the weak function route please provide documentation of why what you are doing is guaranteed to work for JVM and JNI loading.

My request of "use callbacks" comes both from thinking that is clearer code and from being unclear if weak symbols are guaranteed to work the way the PR uses them.

Reply:
The ApplicationInit is not only for init DAC, but also for many others like SetExtendedDiscoveryTimeoutSecs and many others, please ref for examples/all-clusters-app/linux/main-common.cpp and examples/ota-requestor-app/linux/main.cpp

Reply:
> What I am thinking is that java generally has bi-directional support: java can call native code and native code can call java callbacks.
> 
> If the above is correct, which language is what version should not matter:
> 
> * java code can get a native pointer to the "ApplicationInit" method
> * java code can pass in a callback to "startApp" ... maybe implement a thing on "self" that is a callback so you can "getMethod" and call it if exists or some other way.
> 
> I would like the code to make call paths clear: if you need something called at startup, a pointer to such a callback/delegate should be passed in. Compile time seems stranger to me as java dynamically links jni libraries, so I am unclear how attribute weak is guaranteed to behave in a .so environment. I tried to find documentation that explains how these are suposed to work and could not find any - if you go the weak function route please provide documentation of why what you are doing is guaranteed to work for JVM and JNI loading.
> 
> My request of "use callbacks" comes both from thinking that is clearer code and from being unclear if weak symbols are guaranteed to work the way the PR uses them.

Yes, I understand that we can add a java callback to startApp(java to c)  ->  startApp call it (c to java) --> then the callback call native init ( java to c ). I am afraid it is too long. 

And also ApplicationInit is already used in Linux, so I just follow it.